### PR TITLE
Add external dependencies fields to helm remote

### PR DESCRIPTION
**Pull request by tenstad**
Closes https://github.com/jfrog/terraform-provider-artifactory/issues/259

Tested with Artifactory 7.27.3

### CODE REVIEW

**Comment by tenstad** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by tenstad** [comment]
recheck

**Review by chb0github — APPROVED** [review]
Has my approval! Thanks for the PR

File: pkg/artifactory/resource_artifactory_remote_helm_repository.go [hunk 1]
```diff
@@ -13,11 +13,31 @@ var helmRemoteSchema = mergeSchema(baseRemoteSchema, map[string]*schema.Schema{
...
+		Description: "An Allow List of Ant-style path expressions that specify where external dependencies may be downloaded from. " +
+			"By default, this is set to ** which means that dependencies may be downloaded from any external source.",
+	},
 })
 
 type HelmRemoteRepo struct {
 	RemoteRepositoryBaseParams
-	HelmChartsBaseURL string `hcl:"helm_charts_base_url" json:"chartsBaseUrl,omitempty"`
+	HelmChartsBaseURL            string   `hcl:"helm_charts_base_url" json:"chartsBaseUrl,omitempty"`
+	ExternalDependenciesEnabled  bool     `hcl:"external_dependencies_enabled" json:"externalDependenciesEnabled,omitempty"`
```
Comment (alexhung) [inline pkg/artifactory/resource_artifactory_remote_helm_repository.go:39]: Should we continue using `omitempty` here for these new attributes? @chb0github My vote is no but would like to hear your take.

File: pkg/artifactory/resource_artifactory_remote_helm_repository.go [hunk 2]
```diff
@@ -13,11 +13,31 @@ var helmRemoteSchema = mergeSchema(baseRemoteSchema, map[string]*schema.Schema{
...
+	"external_dependencies_patterns": {
+		Type:     schema.TypeList,
+		Optional: true,
+		ForceNew: true,
+		Elem: &schema.Schema{
+			Type: schema.TypeString,
+		},
+		RequiredWith: []string{"external_dependencies_enabled"},
+		Description: "An Allow List of Ant-style path expressions that specify where external dependencies may be downloaded from. " +
+			"By default, this is set to ** which means that dependencies may be downloaded from any external source.",
```
Comment (alexhung) [inline pkg/artifactory/resource_artifactory_remote_helm_repository.go:32]: Should we be explicit about the default here with `Default:  "**",`?

Comment (tenstad) [reply pkg/artifactory/resource_artifactory_remote_helm_repository.go:32]: There is no explicit default in remote docker repo: https://github.com/jfrog/terraform-provider-artifactory/blob/master/pkg/artifactory/resource_artifactory_remote_docker_repository.go#L35-L45

File: pkg/artifactory/resource_artifactory_remote_helm_repository.go (hunk 1 above)
Comment (tenstad) [reply pkg/artifactory/resource_artifactory_remote_helm_repository.go:39]: `GoVirtualRepositoryParams` has `omitempty`, `DockerRemoteRepository` has not.

Comment (tenstad) [reply pkg/artifactory/resource_artifactory_remote_helm_repository.go:39]: Aha, if `omitempty` is removed from `external_dependencies_enabled`, it does not need `ForceNew`. I agree that it should be removed.

Comment (alexhung) [reply pkg/artifactory/resource_artifactory_remote_helm_repository.go]: Yeah, `omitempty` has been a constant/consistent troublemaker for us. Hence I'd like to not add this if at all possible. This should apply to `external_dependencies_patterns` as well.

**Review by alexhung — DISMISSED** [review]
:shipit:

**Comment by tenstad** [comment]
Thanks for the quick review @alexhung @chb0github!

Made a final adjustment to make the example so that is matches the one in https://www.jfrog.com/confluence/display/JFROG/Kubernetes+Helm+Chart+Repositories

File: pkg/artifactory/resource_artifactory_remote_helm_repository.go (hunk 2 above)
Comment (chb0github) [reply pkg/artifactory/resource_artifactory_remote_helm_repository.go:31]: > Should we be explicit about the default here with `Default: "**",`?
I think so.

File: pkg/artifactory/resource_artifactory_remote_helm_repository.go (hunk 1 above)
Comment (chb0github) [reply pkg/artifactory/resource_artifactory_remote_helm_repository.go]: @tenstad - Our experience, over-and-over, has been that `omitempty` is generally bad. It allows the server to interpret its absence as it sees fit, and that behavior is inconsistent and opaque; varying from endpoint-to-endpoint as a given developer decided to handle it. 

Compounding the problem is what go considers `empty` - `0` for int is `empty`, but many times it's a perfectly valid value. Some with `""` blank strings. 

It's not a blanket ban, but using `omitempty` should be used on a case-by-case basis for specific needs. We currently have at least 3 tickets open relating to this

Please remove `omitempty` and put proper default values in those fields so it's clear what the expectations are

File: pkg/artifactory/resource_artifactory_remote_helm_repository.go [hunk 3]
```diff
@@ -36,8 +35,8 @@ var helmRemoteSchema = mergeSchema(baseRemoteSchema, map[string]*schema.Schema{
 type HelmRemoteRepo struct {
 	RemoteRepositoryBaseParams
 	HelmChartsBaseURL            string   `hcl:"helm_charts_base_url" json:"chartsBaseUrl,omitempty"`
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_remote_helm_repository.go:37]: I know it isn't part of your changes, but I think `omitempty` should be removed from here too. @alexhung what do you think?

File: pkg/artifactory/resource_artifactory_remote_helm_repository.go (hunk 2 above)
Comment (tenstad) [reply pkg/artifactory/resource_artifactory_remote_helm_repository.go:31]: I'm not sure how to add a default for a list, tried `Default:  []string{"**"},`.

> 1 error occurred:
>       * resource artifactory_remote_helm_repository: external_dependencies_patterns: Default is not valid for lists or sets

File: pkg/artifactory/resource_artifactory_remote_helm_repository.go (hunk 1 above)
Comment (tenstad) [reply pkg/artifactory/resource_artifactory_remote_helm_repository.go]: We all agree, `omitempty` is removed :smile:

File: pkg/artifactory/resource_artifactory_remote_helm_repository.go (hunk 3 above)
Comment (alexhung) [reply pkg/artifactory/resource_artifactory_remote_helm_repository.go:37]: I agree we should remove this too, assuming tests all passed. If test(s) fails, then let's open a separate issue to for this. I don't want scope creep.

Comment (tenstad) [reply pkg/artifactory/resource_artifactory_remote_helm_repository.go]: Removed it! Could one of you run the tests? I have only been testing manually by modifying `sample.tf`.

Comment (tenstad) [reply pkg/artifactory/resource_artifactory_remote_helm_repository.go]: As the field is required and cannot be `""`, I do not think `omitempty` has any effect here.

File: pkg/artifactory/resource_artifactory_remote_helm_repository.go (hunk 2 above)
Comment (alexhung) [reply pkg/artifactory/resource_artifactory_remote_helm_repository.go:31]: Thanks TF!

[Artifactory schema documentation](https://www.jfrog.com/confluence/display/JFROG/Repository+Configuration+JSON#RepositoryConfigurationJSON-RemoteRepository) has no default value for `externalDependenciesPatterns` so is "**" the implicit default behavior from the API?

File: pkg/artifactory/resource_artifactory_remote_helm_repository.go (hunk 3 above)
Comment (alexhung) [reply pkg/artifactory/resource_artifactory_remote_helm_repository.go]: @tenstad The commit passed our internal CI so this is good to go.

**Review by alexhung — APPROVED** [review]
With the exception on the default value for `externalDependenciesPatterns`, this PR is good to go for me.

Since TF doesn't allow default value to be set for `TypeList` or `TypeSet`, the only question I have is why `**` is mentioned in the original description. This suggests either an implicit behavior of Artifactory (or the API), or a mistake in the description.

**Comment by alexhung** [comment]
Having said that, I just read the [doc](https://www.jfrog.com/confluence/display/JFROG/Kubernetes+Helm+Chart+Repositories#KubernetesHelmChartRepositories-RemoteRepositories) and it clearly states the `**` default.

**Comment by tenstad** [comment]
I have been using https://github.com/tenstad/terraform-provider-artifactory/blob/tenstad/pkg/artifactory/resource_artifactory_remote_repository.go#L210-L215 for a while and checked my Artifactory insance. Found a lot of helm remotes with `externalDependenciesPatterns: ["**"]`, where no value is configured in terraform.

When I now added a new helm remote with terraform (with the provider linked above), it got `externalDependenciesPatterns: []`. I think the default has been changed from `["**"]` to `[]` in a semi-recent Artifactory release. However, when adding a remote helm repo in GUI, the default is still `["**"]`.

@alexhung should we update the description, and remove `By default, this is set to ** which means that dependencies may be downloaded from any external source.`?

Sorry for not discovering this earlier, I have been used to the `["**"]` default, and as you say, the documentation also says so.

**Comment by alexhung** [comment]
@tenstad I can confirm the API defaults the `externalDependenciesPatterns` attribute to `[]`. So let's update the attribute description to mention this discrepancy between the UI and API.

**Pull request merged**